### PR TITLE
Hide secrets config in console log

### PR DESCRIPTION
**Pull request by mingwli**
[AB#406637](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/406637)

### CODE REVIEW

File: src/docfx/config/ConfigLoader.cs [hunk 1]
```diff
@@ -61,7 +61,8 @@ where glob(file)
             var cliConfig = new JObject();
             JsonUtility.Merge(unionProperties, cliConfig, options.StdinConfig, options.ToJObject());
 
-            // TODO: print stdin passed secrets
+            Console.WriteLine($"stdin secret config: " +
```
Comment (yufeih) [inline src/docfx/config/ConfigLoader.cs:64]: Use `Log.Write`

File: src/docfx/lib/MaskUtility.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Linq;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class MaskUtility
```
Comment (yufeih) [inline src/docfx/lib/MaskUtility.cs:9]: move to `StringUtility`?

File: test/docfx.Test/lib/MaskUtilityTest.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+using Xunit;
+
+namespace Microsoft.Docs.Build
+{
+    public static class MaskUtilityTest
+    {
+        [Fact]
+        public static void HideSecret()
+        {
+            var githubToken = TestUtility.RandomSHA1Hash();
```
Comment (yufeih) [inline test/docfx.Test/lib/MaskUtilityTest.cs:16]: why this `RandomSHA1Hash` method? Can `Guid.NewGuid().ToString()` work?

File: src/docfx/config/ConfigLoader.cs [hunk 4]
```diff
@@ -61,7 +61,8 @@ where glob(file)
...
+                $"{JsonUtility.Serialize(MaskUtility.HideSecret(options?.StdinConfig?["secrets"] ?? new JObject()), true)}");
```
Comment (yufeih) [inline src/docfx/config/ConfigLoader.cs:65]: nit: `JToken.ToString()` would give you a formatted JSON output.

Comment (yufeih) [inline src/docfx/config/ConfigLoader.cs:65]: This method does not show other stdin properties

File: test/docfx.Test/lib/MaskUtilityTest.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                    {
+                        "https://buildapi.docs.microsoft.com",
+                        new { headers = new Dictionary<string, string> { ["X-OP-BuildUserToken"] = opBuildUserToken } }
+                    },
+                },
+                arr = secrets,
+            };
+
+            var serialized = JsonUtility.Serialize(MaskUtility.HideSecret(JsonUtility.ToJObject(data)), indent: true);
+            Assert.True(secrets.All(secret => !serialized.Contains(secret)));
```
Comment (yufeih) [inline test/docfx.Test/lib/MaskUtilityTest.cs:37]: assert serialized contains `***`

File: test/docfx.Test/lib/MaskUtilityTest.cs (hunk 3 above)
Comment (mingwli) [reply test/docfx.Test/lib/MaskUtilityTest.cs:16]: I checked that github PAT is a string of 40 chars with [0-9a-f], more like a SHA1 hash

File: src/docfx/lib/MaskUtility.cs (hunk 2 above)
Comment (mingwli) [reply src/docfx/lib/MaskUtility.cs:9]: It also handles JToken, mask comes more reasonable

File: test/docfx.Test/lib/MaskUtilityTest.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            Assert.Contains("***", serialized);
```
Comment (herohua) [inline test/docfx.Test/lib/MaskUtilityTest.cs:38]: Shall we have stricter checks? Current check cannot catch the bug if the secret is printed as `{secret.Substring(0, 10)}***{secret.Substring(10)}`

File: test/docfx.Test/TestUtility.cs [hunk 7]
```diff
@@ -188,6 +189,20 @@ public static JToken ApplyVariables(JToken value, IEnumerable<KeyValuePair<strin
             return value;
         }
 
+        public static string RandomSHA1Hash()
+        {
```
Comment (herohua) [inline test/docfx.Test/TestUtility.cs:193]: don't use SHA1 even for non-crypto scenarios - this will raise red flag in security reportings

**Review by herohua — CHANGES_REQUESTED** [review]
:clock1:

File: test/docfx.Test/lib/MaskUtilityTest.cs (hunk 6 above)
Comment (mingwli) [reply test/docfx.Test/lib/MaskUtilityTest.cs:38]: I'm not following, when may `{secret.Substring(10)}` occurs?

File: test/docfx.Test/TestUtility.cs (hunk 7 above)
Comment (mingwli) [reply test/docfx.Test/TestUtility.cs:193]: lemme figure out another way to generate a 40 digits hex string

Comment (mingwli) [reply test/docfx.Test/TestUtility.cs:193]: using random 40 digits hex string

**Comment by mingwli** [comment]
image
verified

File: test/docfx.Test/TestUtility.cs [hunk 8]
```diff
@@ -8,6 +8,7 @@
 using System.IO.Compression;
 using System.Linq;
 using System.Reflection;
+using System.Security.Cryptography;
```
Comment (yufeih) [inline test/docfx.Test/TestUtility.cs:11]: revert

**Review by yufeih — APPROVED** [review]

File: test/docfx.Test/lib/MaskUtilityTest.cs (hunk 6 above)
Comment (herohua) [reply test/docfx.Test/lib/MaskUtilityTest.cs:38]: As discussed offline, please re-consider my suggestion

Comment (mingwli) [reply test/docfx.Test/lib/MaskUtilityTest.cs:38]: make sense!

**Review by herohua — APPROVED** [review]
:shipit:

**Pull request merged**